### PR TITLE
[release/v8.2] Fix multiple external libraries import

### PR DESCRIPTION
**Pull request by chetan-zilliqa**
## Description

Here is the test plan executed for the external library support feature.
https://docs.google.com/spreadsheets/d/1bL1LwMXVlctcyI7RgfzISkvv050K9Wss0x85m8dTs_I/edit#gid=1082855032

## Backward Compatibility

- [ ] This is not a breaking change
- [ ] This is a breaking change

### Implementation

- [x] **ready for review**

### Integration Test (Core Team)

- [ ] local machine test

- [x] small-scale cloud test

### CODE REVIEW

File: src/libData/AccountData/AccountStoreSC.cpp [hunk 1]
```diff
@@ -540,9 +538,11 @@ bool AccountStoreSC<MAP>::UpdateAccounts(const uint64_t& blockNum,
       }
 
       if (is_library) {
+        // Call to library should silently passed to scilla so that the gas is
+        // charged based on the size of the message
         LOG_GENERAL(WARNING, "Library being called");
-        error_code = TxnStatus::FAIL_SCILLA_LIB;
-        return false;
+        // error_code = TxnStatus::FAIL_SCILLA_LIB;
+        // return false;
```
Comment (jjcnn) [inline src/libData/AccountData/AccountStoreSC.cpp:545]:
```suggestion
```

File: src/libData/AccountData/AccountStoreSC.cpp [hunk 2]
```diff
@@ -1449,9 +1464,11 @@ bool AccountStoreSC<MAP>::ParseCallContractJsonOutput(
       }
 
       if (is_library) {
+        // Scilla should be invoked for message sent to library so that the GAS
+        // is charged
         LOG_GENERAL(WARNING, "Library being called");
-        receipt.AddError(LIBRARY_AS_RECIPIENT);
-        return false;
+        // receipt.AddError(LIBRARY_AS_RECIPIENT);
+        // return false;
```
Comment (jjcnn) [inline src/libData/AccountData/AccountStoreSC.cpp:1471]:
```suggestion
```

File: docker/Dockerfile [hunk 3]
```diff
@@ -14,7 +14,7 @@
 # along with this program.  If not, see <https://www.gnu.org/licenses/>.
 #
 
-ARG BASE=zilliqa/scilla:v0.12.0rc1
+ARG BASE=zilliqa/scilla:a8d0c412
```
Comment (jjcnn) [inline docker/Dockerfile:17]: Should this line be merged? The commit that this refers to is now merged to `master`.

**Review by jjcnn — COMMENTED** [review]
This looks ok to my untrained eyes, but somebody else needs to do a proper review. I can only really review things as they relate to the Scilla logic.

File: docker/Dockerfile (hunk 3 above)
Comment (chetan-zilliqa) [reply docker/Dockerfile:17]: I will revert this commit before the merge with the new scilla tag. Only kept it for testing.

File: src/libData/AccountData/AccountStoreSC.cpp (hunk 2 above)
Comment (chetan-zilliqa) [reply src/libData/AccountData/AccountStoreSC.cpp:1471]: Clang-format tools takes care of it automatically.

File: src/libData/AccountData/AccountStoreSC.cpp (hunk 1 above)
Comment (chetan-zilliqa) [reply src/libData/AccountData/AccountStoreSC.cpp:545]: Clang-format tools takes care of it automatically.

File: src/libData/AccountData/Account.cpp [hunk 4]
```diff
@@ -135,7 +135,13 @@ void AccountBase::SetCodeHash(const dev::h256& codeHash) {
 
 const dev::h256& AccountBase::GetCodeHash() const { return m_codeHash; }
 
-bool AccountBase::isContract() const { return m_codeHash != dev::h256(); }
+bool Account::isContract() const {
+  return (m_codeHash != dev::h256() && !m_is_library);
```
Comment (yaron-zilliqa) [inline src/libData/AccountData/Account.cpp:139]: A small suggestion: better to put the "cheap" condition first:
`return  !m_is_library && m_codeHash != dev::h256()`

File: src/libData/AccountData/Account.cpp [hunk 5]
```diff
@@ -135,7 +135,13 @@ void AccountBase::SetCodeHash(const dev::h256& codeHash) {
...
+}
+
+bool Account::IsLibrary() const {
+  return (m_codeHash != dev::h256() && m_is_library);
```
Comment (yaron-zilliqa) [inline src/libData/AccountData/Account.cpp:143]: As above.

**Review by yaron-zilliqa — CHANGES_REQUESTED** [review]

File: src/libUtils/ScillaUtils.cpp [hunk 6]
```diff
@@ -108,8 +109,13 @@ Json::Value ScillaUtils::GetCallContractJson(const string& root_w_version,
   ret["argv"].append(boost::filesystem::current_path().string() + '/' +
                      OUTPUT_JSON);
   ret["argv"].append("-i");
-  ret["argv"].append(boost::filesystem::current_path().string() + '/' +
-                     INPUT_CODE + CONTRACT_FILE_EXTENSION);
+  if (is_library) {
```
[the remark this answers is not in the document]
Comment (chetan-zilliqa) [reply src/libUtils/ScillaUtils.cpp:112]: I got your point about changing code like this.
```ret["argv"].append((
+      boost::filesystem::current_path() /  INPUT_CODE  / (is_library ? LIBRARY_CODE_EXTENSION : CONTRACT_FILE_EXTENSION)).string());```

However for consistency. I feel that let's go with your first suggestion and change them later on with a code refactoring ticket.

https://zilliqa-jira.atlassian.net/browse/ZIL-4891

File: src/libData/AccountData/Account.cpp (hunk 4 above)
Comment (chetan-zilliqa) [reply src/libData/AccountData/Account.cpp:139]: done

File: src/libData/AccountData/Account.cpp (hunk 5 above)
Comment (chetan-zilliqa) [reply src/libData/AccountData/Account.cpp:143]: done

File: src/libUtils/ScillaUtils.h [hunk 7]
```diff
@@ -41,7 +41,7 @@ class ScillaUtils {
   /// get the command for invoking the scilla_runner while calling
   static Json::Value GetCallContractJson(
       const std::string& root_w_version, const uint64_t& available_gas,
-      const boost::multiprecision::uint128_t& balance);
+      const boost::multiprecision::uint128_t& balance, const bool& is_library);
```
[the remark this answers is not in the document]
Comment (chetan-zilliqa) [reply src/libUtils/ScillaUtils.h:44]: It should be ok as it's a constant reference—no need to create an additional local variable. Right?

File: src/libData/AccountData/AccountStoreSC.cpp [hunk 8]
```diff
@@ -863,24 +862,37 @@ bool AccountStoreSC<MAP>::ExportContractFiles(
...
   }
+  if (ENABLE_CHECK_PERFORMANCE_LOG) {
+    LOG_GENERAL(INFO, "LDB Read (microsec) = " << r_timer_end(tpStart));
+  }
 
   return true;
 }
 
+template <class MAP>
+void AccountStoreSC<MAP>::CreateScillaCodeFiles(
```
[the remark this answers is not in the document]
Comment (chetan-zilliqa) [reply src/libData/AccountData/AccountStoreSC.cpp:882]: I only moved the common code into the function that consumes both  scilla library/contract. I believe the original purpose could be not to increase header file size(compilation time) or something like that.

**Comment by chetan-zilliqa** [comment]
This will redone on release/v8.3 branch

**Pull request closed without merging**